arangod/Network/NetworkFeature.cpp

```diff
@@ -23,30 +23,232 @@
+
+    _mutex.lock();
+    _mutex.unlock();
+    // notify retry thread
+    _cv.notify_one();
+  }
+
+  void cancelAll() noexcept {
+    std::unique_lock<std::mutex> guard(_mutex);
+
+    // pop everything from the queue until it is empty.
+    while (!_retryRequests.empty()) {
+      auto& top = _retryRequests.top();
+      auto req = std::move(top.req);
+      TRI_ASSERT(req);
+      _retryRequests.pop();
+      try {
+        // canceling a request can throw in case a concurrent
+        // thread has already resolved or canceled the request.
+        // in this case, simply ignore the exception.
+        req->cancel();
+      } catch (...) {
+        // does not matter
+      }
+    }
+  }
+
+  void push(std::shared_ptr<network::RetryableRequest> req,
+            std::chrono::steady_clock::time_point retryTime) {
+    TRI_ASSERT(req);
+
+    auto cancelGuard = scopeGuard([req]() noexcept {
+      try {
+        // canceling a request can throw in case a concurrent
+        // thread has already resolved or canceled the request.
+        // in this case, simply ignore the exception.
+        req->cancel();
+      } catch (...) {
+        // does not matter
+      }
+    });
+
+    TRI_IF_FAILURE("NetworkFeature::retryRequestFail") {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+    }
+
+    std::unique_lock<std::mutex> guard(_mutex);
+
+    if (isStopping()) {
+      return;
+    }
+
+    bool mustNotify = (retryTime < _nextRetryTime);
+    if (mustNotify) {
+      _nextRetryTime = retryTime;
+    }
+
+    _retryRequests.push(RetryItem{retryTime, std::move(req)});
+    guard.unlock();
+
+    cancelGuard.cancel();
+
+    if (!mustNotify) {
+      // retry time is already in the past?
+      mustNotify = retryTime <= std::chrono::steady_clock::now();
+    }
+
+    // notify retry thread about the new item
+    if (mustNotify) {
+      _cv.notify_one();
+    }
+  }
+
+ protected:
+  void run() override {
+    while (!isStopping()) {
+      try {
+        std::unique_lock<std::mutex> guard(_mutex);
+
+        // by default, sleep an arbitrary amount of 10 seconds.
+        // note that this value may be reduced if we have an
+        // element in the queue that is due earlier.
+        _nextRetryTime = std::chrono::steady_clock::now() + kDefaultSleepTime;
+
+        while (!_retryRequests.empty()) {
+          auto& top = _retryRequests.top();
+          auto now = std::chrono::steady_clock::now();
+          if (top.retryTime > now) {
+            // next retry operation is in the future...
+            _nextRetryTime = top.retryTime;
+            break;
+          }
+
+          auto req = std::move(top.req);
+          TRI_ASSERT(req);
+          _retryRequests.pop();
+
+          if (!_retryRequests.empty()) {
+            _nextRetryTime = _retryRequests.top().retryTime;
+          } else {
+            _nextRetryTime = now + kDefaultSleepTime;
+          }
+
+          guard.unlock();
+
+          if (isStopping()) {
+            break;
```

Comment:
If we hit this, we would still go into the `wait_until`, so this could be a large delay, if the next retry is far in the future. Maybe we want to `return` here directly?

Reply:
`return`ing is not sufficient, because we also want to call `cancelAll()` once. Adjusted the code now to break the loop properly without waiting afterwards.

Reply:
Should we call `req->cancel()` here? Otherwise this request will just be dropped, won't it?

Reply:
OK, the `break` is good, it leads to termination of the outer loop, too and thus to `cancelAll`, agreed. But Tobias comment is also good, because `req` is already off the retry queue...